### PR TITLE
Internal workflow engine feature flag

### PR DESCRIPTION
**Pull request by johnewart**
# Description

Adds a feature flag for workflow engine to ensure that it doesn't get enabled unless explicitly turned on.

## Issue reference

Part of the alpha release of workflows tracked in #5537

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing

### CODE REVIEW

**Comment by yaron2** [comment]
We don't require a feature flag for alpha building blocks. Case in point, Configuration API and Distributed Lock.

**Comment by ItalyPaleAle** [comment]
Wait, this is not a feature flag but rather just a CLI flag.

Why do we need a CLI flag to enable this? Does it alter the behavior of the runtime in some ways that it could cause conflicts with other parts of Dapr if it's enabled?

File: pkg/actors/internal/placement.go [hunk 1]
```diff
@@ -112,6 +113,18 @@ func NewActorPlacement(
 	}
 }
 
+// Force send an update to placement service (i.e late registration of internal actors)
```
Comment (cgillum) [inline pkg/actors/internal/placement.go:116]: Q: Which part of the implementation forces the update to the placement service?

**Review by cgillum — APPROVED** [review]
LGTM. Just had one question related to a comment in placement.go.

File: pkg/actors/internal/placement.go (hunk 1 above)
Comment (johnewart) [reply pkg/actors/internal/placement.go:116]: I will change that, it doesn't actually force an update, the update will happen with the heartbeat

File: pkg/runtime/cli.go [hunk 2]
```diff
@@ -363,6 +365,11 @@ func FromFlags() (*DaprRuntime, error) {
 		globalConfig = daprGlobalConfig.LoadDefaultConfiguration()
 	}
 
+	if !daprGlobalConfig.IsFeatureEnabled(globalConfig.Spec.Features, daprGlobalConfig.WorkflowEngine) && *enableWorkflowEngine {
```
Comment (yaron2) [inline pkg/runtime/cli.go:368]: @ItalyPaleAle your comment is correct about the CLI flag, but it seems to be that AND the feature preview.

**Comment by johnewart** [comment]
> Why do we need a CLI flag to enable this? Does it alter the behavior of the runtime in some ways that it could cause conflicts with other parts of Dapr if it's enabled?

Since the engine will create and register internal actors that run constantly and do work with this enabled, the intention was to require that people turn it on (and you are right it looks like I forgot a change that checks `daprGlobalConfig` -- I must have stashed it while doing something else, I just pushed it after the tests finished). While the chances are low, it would mean that the blast radius of a bug in the workflow engine itself would be limited to those who explicitly opt-in to . 

Also I just realized the title should explicitly call out that it's for the _engine_, not the building block (also re: @yaron2's comment)

**Comment by yaron2** [comment]
> Wait, this is not a feature flag but rather just a CLI flag.
> 
> Why do we need a CLI flag to enable this? Does it alter the behavior of the runtime in some ways that it could cause conflicts with other parts of Dapr if it's enabled?

Very good questions indeed. From what I observe, since Workflow Engine uses actors, then the runtime will try to initialize actors (and look for a compatible state store) by default unless instructed otherwise.

**Comment by johnewart** [comment]
Changed the title to reflect that the engine is the part in question, not the building block itself

**Comment by yaron2** [comment]
> Also I just realized the title should explicitly call out that it's for the engine, not the building block (also re: @yaron2's comment)

Thanks, but I'm not sure we want to separate the built-in Workflows from the management APIs. Essentially they both fall under the Workflows building block. Need to give this some thought.

**Comment by johnewart** [comment]
The management API can be used, for example, to manage Temporal workflows so it could exist without the internal engine. As an alternative we can change it so that the workflow engine only gets started if a built-in workflow component exists, which is mostly the same logic as here only we execute it when the workflow component is initialized instead of based on a flag.

**Comment by yaron2** [comment]
> As an alternative we can change it so that the workflow engine only gets started if a built-in workflow component exists

This sounds like a much better approach.

**Comment by johnewart** [comment]
Sounds good, I will update the PR with the component-registration bits as well

**Comment by cgillum** [comment]
> As an alternative we can change it so that the workflow engine only gets started if a built-in workflow component exists

I'd prefer we not go this route. The current design allows users to use the workflow engine even without using registering an API component. For example, a Durable Task SDK can interact with workflows directly on the workflow gRPC endpoint (this is how all my demos work today since we don't yet have the built-in API component). I'd hate to take that capability away if there isn't a really good reason to do so.  My thinking is that we keep the building block API and the engine decoupled.

**Comment by yaron2** [comment]
> > As an alternative we can change it so that the workflow engine only gets started if a built-in workflow component exists
> 
> I'd prefer we not go this route. The current design allows users to use the workflow engine even without using registering an API component. For example, a Durable Task SDK can interact with workflows directly on the workflow gRPC endpoint (this is how all my demos work today since we don't yet have the built-in API component). I'd hate to take that capability away if there isn't a really good reason to do so. My thinking is that we keep the building block API and the engine decoupled.

I'm sure I'm missing something then - just so we're all clear: what would a "built-in workflow component" enable that using the workflow engine without an API component doesn't?

**Pull request closed without merging**